### PR TITLE
discard useless cards

### PR DESCRIPTION
**Pull request by fpvandoorn**
This is the first sml code I've ever written, so any suggestions about coding style, standard function I'm not using when appropriate or cringe-worthy code are welcome!

One particular question:
- Are there functions we can use other than the [SML basis library](http://sml-family.org/Basis/) and anything in `cmlib`? In particular, where is `Util.revFind`? Did I reimplement some existing functions (`mapFilter` or `hdOption`)?

### CODE REVIEW

**Comment by fpvandoorn** [comment]
Oh, I just found `util.sml`. I'll move some things there.

**Comment by cangiuli** [comment]
Why did you add `.cm` to the `.gitignore`?

Can you update this with the new code we discussed, and perhaps clean up the commits (with rebase) to get rid of the intermediate versions of things?

**Comment by cangiuli** [comment]
By the way, the function checking whether something is in a list should be called `elem`.

**Comment by fpvandoorn** [comment]
> Why did you add `.cm` to the `.gitignore`?

When I run `rlwrap sml -m sources.cm` I get a folder `src/.cm` which then is listed in `git status`. `.gitignore` is to make sure such files/folders don't show up in `git status`, right?

> Can you update this with the new code we discussed, and perhaps clean up the commits (with rebase) to get rid of the intermediate versions of things?
> By the way, the function checking whether something is in a list should be called `elem`.

I will do/change these things.

**Comment by cangiuli** [comment]
Oh, right, I was thinking you wrote `*.cm`.

**Comment by fpvandoorn** [comment]
Ok, the pull request is ready to go, in my opinion. Feel free to change anything else in it.

**Comment by fpvandoorn** [comment]
Ok, the pull request is ready to go now. Feel free to make any other changes, though.

File: src/simple.sml [hunk 1]
```diff
@@ -30,7 +31,7 @@ struct
 
   (* Returns index of oldest unclued card in our own hand. *)
   fun ourOldestUnclued (xs : info list list) : int option =
-    oldestUnclued (ListPair.zip (List.tabulate (5,fn n => n), xs))
+    Util.revFindIndex (fn is => not (isClued is)) xs
```
Comment (cangiuli) [inline src/simple.sml]: FYI you can also write the first argument `not o isClued`.

File: src/simple.sml [hunk 2]
```diff
@@ -50,6 +51,65 @@ struct
...
+  (* Gives the list of negative rank information of a card *)
+  fun cluedNotRank (is : info list) : int list =
+    List.mapPartial (fn i => case i of NotRank r => SOME r | _ => NONE) is
+
+  (* Gives the list of negative color information of a card *)
+  fun cluedNotSuit (is : info list) : suit list =
+    List.mapPartial (fn i => case i of NotSuit su => SOME su | _ => NONE) is
+
+  (* Returns whether a card is clued to be not rainbow *)
+  fun isNotRainbow (is : info list) : bool =
```
Comment (cangiuli) [inline src/simple.sml]: Isn't this just `not (null (cluedNotSuit is))`? I think you can just inline that inside of `possibleSuits`.

File: src/simple.sml [hunk 3]
```diff
@@ -50,6 +51,65 @@ struct
...
+  (* Gives the clued suit, returns Rainbow if 2 different suit clues have been given *)
+  fun cluedSuit (is : info list) : suit option =
+  let
+    val sus = List.mapPartial (fn i => case i of IsSuit su => SOME su | _ => NONE) is
+  in
+    if null sus then NONE
+    else if List.exists (fn j => j <> hd sus) (tl sus)
+    then SOME Rainbow
+    else SOME (hd sus)
+  end
```
Comment (cangiuli) [inline src/simple.sml]:
```
  fun cluedSuit (is : info list) : suit option =
    case List.mapPartial (fn i => case i of IsSuit su => SOME su | _ => NONE) is of
         [] => NONE
       | x::xs => if List.exists (fn su => su <> x) xs then SOME Rainbow else SOME x
```

File: src/util.sml [hunk 4]
```diff
@@ -50,4 +57,12 @@ struct
          SOME v => v
        | NONE => f ()
 
+  (* The head of the list which returns an option type *)
+  fun hdOption (l : 'a list) : 'a option =
+    case l of nil => NONE | x :: l' => SOME x
```
Comment (cangiuli) [inline src/util.sml]: I think this isn't needed anywhere anymore, right?

File: src/simple.sml [hunk 5]
```diff
@@ -50,6 +51,65 @@ struct
...
+  fun possibleSuits (is : info list) : suit list =
+    case cluedSuit is of
+	 SOME Rainbow => [Rainbow]
+       | SOME su => if isNotRainbow is then [su] else [su, Rainbow]
+       | NONE =>
+	 let
+	     val lsu = cluedNotSuit is
+	 in
+	     (List.filter (fn su => List.all (fn su' => su <> su') lsu)
+			  normalSuits) @ (if null lsu then [Rainbow] else [])
```
Comment (cangiuli) [inline src/simple.sml]: I think the first part of this is `List.filter (fn su => not (Util.elem lsu su)) normalSuits`.

Comment (cangiuli) [reply src/simple.sml]: Or `List.filter (not o (Util.elem lsu)) normalSuits`.

**Comment by fpvandoorn** [comment]
updated

**Comment by cangiuli** [comment]
Great, thanks! Any thoughts on #4?

**Pull request merged**